experimental/dds/tree2/src/simple-tree/factoryInputTypes.ts

```diff
@@ -0,0 +1,110 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { TreeValue } from "../core";
+import { RestrictiveReadonlyRecord } from "../util";
+import {
+	FieldKind,
+	FieldKinds,
+	AllowedTypes,
+	Any,
+	FieldNodeSchema,
+	TreeFieldSchema,
+	InternalTypedSchemaTypes,
+	LeafNodeSchema,
+	MapNodeSchema,
+	ObjectNodeSchema,
+	TreeNodeSchema,
+	TreeSchema,
+} from "../feature-libraries";
+
+/**
+ * Data from which a {@link TreeObjectNode} can be built.
+ * @alpha
+ */
+export type TreeObjectNodeFactoryInput<TSchema extends ObjectNodeSchema> =
+	TreeObjectNodeFieldsFactoryInput<TSchema["objectNodeFieldsObject"]>;
+
+/**
+ * Helper for generating the properties of a TreeObjectNodeFactoryInput.
+ * @alpha
+ */
+export type TreeObjectNodeFieldsFactoryInput<
+	TFields extends RestrictiveReadonlyRecord<string, TreeFieldSchema>,
+> = {
+	// Make all properties optional.
+	readonly [key in keyof TFields]?: TreeFieldFactoryInput<TFields[key]>;
+} & {
+	// Require non-optional.
+	readonly [key in keyof TFields as TFields[key]["kind"] extends typeof FieldKinds.optional
+		? never
+		: key]-?: TreeFieldFactoryInput<TFields[key]>;
+};
+
+/**
+ * Data from which a {@link TreeField} can be built.
+ * @alpha
+ */
+export type TreeFieldFactoryInput<TSchema extends TreeFieldSchema = TreeFieldSchema> =
+	TreeFieldInnerFactoryInput<TSchema["kind"], TSchema["allowedTypes"]>;
+
+/**
+ * Helper for implementing TreeFieldFactoryInput.
```

Comment:
Nit
```suggestion
 * Helper for implementing {@link TreeFieldFactoryInput}.
```

Reply:
Thet will break API extractor since this type is exported via the internal scope. I don't think there is a way to make that link work with API extractor since I think it doesn't support import paths in links.

Reply:
Do these need to be exported then? It seems incorrect to document something that isn't internal in terms of something that is.

Reply:
Implementation details of type meta functions which should never be referred to by users of th epackage have to be exported. When we have such a type, but do not want to guarantee it will continue to exist in its current form (ex: its just an implementation of other APIs which are stable), we export them under "Internal"

Reply:
Note that this helper, which is an implementation detail of the thing its trying to link is inherently less public than the thing its trying to link. Neither the type being referenced or the type referencing it should be consider stable APIs.

Reply:
Oh, I misunderstood the issue. Once we're using api-extractor to manage `/internal` exports, this shouldn't be a problem.